### PR TITLE
Add guideline for cast() and retype().

### PR DESCRIPTION
**Pull request by munificent**

### CODE REVIEW

File: src/_guides/language/effective-dart/usage.md [hunk 1]
```diff
@@ -441,11 +441,41 @@ var reciprocals = stuff.map((n) => 1 / n).cast<double>();
...
+This is the softer generalization of the previous rule. Sometimes there is no
+nearby operation you can use to fix the type of some object. Even then, when
+possible avoid using `cast()` or `retype()` to "change" a collection's type.
+
+Other options that are usually better are:
+
+*   Change the code where the collection is first created so that it has the
+    right type.
+
+*   If you are immediately iterating over the collection, cast each element
```
Comment (kwalrath) [inline src/_guides/language/effective-dart/usage.md]: you are immediately iterating -> you immediately iterate

File: src/_guides/language/effective-dart/usage.md [hunk 2]
```diff
@@ -441,11 +441,41 @@ var reciprocals = stuff.map((n) => 1 / n).cast<double>();
...
+    inside the iteration.
+
+*   If you will eventually access most of the elements in the collection, and
+    you don't need the object to be backed by the original live object, convert
+    it eagerly to the desired type using `List.from()`. `cast()` and `retype()`
```
Comment (kwalrath) [inline src/_guides/language/effective-dart/usage.md]: This third paragraph has some issues (complicated sentences, content that might be relevant outside of the paragraph, sentences that start with a lowercase letter, ...). How about:

-----------

* If you'll...`List.from()`.

These alternatives... careful.

The `cast()` and `retype()` methods return lazy collections that check the element type on _every operation._ If you perform only a few operations on only a few elements, that laziness can be good. But in many cases, the overhead of lazy validation and of wrapping outweighs the benefits.

-----------

If that last paragraph really is relevant only to the 3rd item, then just make it a separate paragraph in the 3rd item.

File: src/_guides/language/effective-dart/usage.md [hunk 3]
```diff
@@ -441,11 +441,41 @@ var reciprocals = stuff.map((n) => 1 / n).cast<double>();
 {% endprettify %}
 
 
+## AVOID using `cast()` or `retype()`.
```
[the remark this answers is not in the document]
Comment (munificent) [reply src/_guides/language/effective-dart/usage.md]: Oops, done.

File: src/_guides/language/effective-dart/usage.md [hunk 4]
```diff
@@ -441,11 +441,41 @@ var reciprocals = stuff.map((n) => 1 / n).cast<double>();
...
+
+This is the softer generalization of the previous rule. Sometimes there is no
+nearby operation you can use to fix the type of some object. Even then, when
+possible avoid using `cast()` or `retype()` to "change" a collection's type.
+
+Other options that are usually better are:
```
[the remark this answers is not in the document]
Comment (munificent) [reply src/_guides/language/effective-dart/usage.md]: Used "prefer", but done.

File: src/_guides/language/effective-dart/usage.md (hunk 1 above)
Comment (munificent) [reply src/_guides/language/effective-dart/usage.md]: Done.

File: src/_guides/language/effective-dart/usage.md (hunk 2 above)
Comment (munificent) [reply src/_guides/language/effective-dart/usage.md]: So much better! Done.

**Comment by munificent** [comment]
Thanks!

**Pull request merged**